[MRG+1] Custom token processor example

rth opened the pull request:
Rebased version of PR #1735, fixes Issue #1156.

This would allow to plug-in a custom token processor for text feature extraction, which can be used e.g. for stemming.

This PR addresses  review comments by @GaelVaroquaux  in the original PR.

Since this is a relatively demanded feature (cf. comments in the original PR) with the general idea mostly validated in 2013: would it be possible to include this into the 0.18 release?

wrichert committed:
Cleaned and updated version of the token processor (PR #1735)

rth committed:
Token processor: addressing PR review comments

amueller commented:
It looks to me that the outcome of #1156 was a different solution. maybe @larsmans or @ogrisel care to comment?
I would also thing that creating custom analyzers should be made easier. Otherwise we end up with arbitrary many callbacks.
What you want to do is relatively easy to achieve by providing a custom analyzer, right? Or is it the ``n_grams`` that give you trouble?

We could maybe add an example on how to build a custom analyzer that can handle n-grams?

rth commented:
Well,  issue #1156 is marked to be automatically closed (not sure how that works) when PR #1735 is merged, which itself was labeled as "Need contributor" not so long ago by @jnothman (since people keep commenting there). I don't personally have a problem with this issue.

I agree this can be achieved with a custom analyser, but this simplifies the process from [this example]( http://stackoverflow.com/a/36191362) to,
     
    stem_vectorizer = CountVectorizer(token_processor=FrenchStemmer().stem)

though the nested callbacks are indeed not pretty.

jnothman commented:
@rth, you should know that we're working towards a release now and seeing as this is not critical (though admittedly it's long-sought) it probably won't be reviewed in detail right now. Please bump once 0.18 has been released over the coming weeks.

amueller commented:
I'd bump this but I think first we want 0.18.1 out the door ;)

rth committed:
Addressing performance issue

rth commented:
I rebased and changed a bit the code to bypass this new text token processor if it is not specified (the previous version  had a 10% performance cost for Bag of words with default parameters). 

Not so happy that this adds yet another parameter to CountVectorizer, etc. However, fundamentally this reflects the problem that `preprocessor`, `tokenizer` and `token_processor` are somewhat redundant with the custom `analyzer` argument, but at the same time it's not trivial to add processing steps to the analyzer without getting into the internals of how it works. For instance, to add stemming with the current code base, one has to create an analyzer that inherits  from the default analyzer (cf. [post above](https://github.com/scikit-learn/scikit-learn/pull/7286#issuecomment-243284274)) which is somewhat non-standard. Not sure what can be done about it though..

P.S: I know 0.18.1 is still in progress, so nothing critical here..

jnothman commented on sklearn/feature_extraction/text.py:
```diff
@@ -1145,6 +1173,10 @@ class TfidfVectorizer(CountVectorizer):
         Only applies if ``analyzer == 'word'``.
 
+    token_processor: callable or None (default)
+        Override the token processing step while preserving the
```
Needs to be clear that it applies to a sequence of tokens, rather than each token individually.

jnothman commented on sklearn/feature_extraction/text.py:
```diff
@@ -1145,6 +1173,10 @@ class TfidfVectorizer(CountVectorizer):
         preprocessing and n-grams generation steps.
         Only applies if ``analyzer == 'word'``.
 
+    token_processor: callable or None (default)
```
space before colon.

jnothman reviewed (commented):
Sorry, @rth, I see the need for this example, but not this feature. You're composing an arbitrary function onto another arbitrary function; and it applies only to words.

What's wrong with:

``` python
class StemmingMixin(object):
    def build_tokenizer(self):
        base = super(StemmingMixin, self).build_tokenizer()
        return lambda doc: map(stem, base(doc))
```

?

rth committed:
Converting this PR to be an example

rth commented:
@jnothman Thanks for the feedback and sorry for late response. 

You are right, this PR was far from ideal. So I removed all code changes, and simply kept the example in the docs on how to use a custom token processor (the same could be used for stemming though I'm not sure it's the most elegant example). Issue https://github.com/scikit-learn/scikit-learn/issues/1156 could then be closed with a "wont fix" status (i.e. one has to use a custom analyser / tokenizer for stemming).

This also removes all `min_df=1` arguments to `CountVectorizer` & `TfidfVecorizer` in the user_manual, since it's the default anyway.

cc @agramfort

jnothman commented on doc/modules/feature_extraction.rst:
```diff
@@ -916,6 +916,30 @@ Some tips and tricks:
     (Note that this will not filter out punctuation.)
 
+
+    The following example will, for instance, transform British spelling to American
```
"some"!

jnothman replied:
This is when I say "so much for all those input parameters; overloading ftw!"

I'm sure this kind of thing is done in the wild, so we might as well advertise it.

LGTM!

vene replied:
Very nice example!

Another use for this, maybe more useful in practice, is normalizing all numbers to a new `##` token

vene commented on doc/modules/feature_extraction.rst:
```diff
@@ -916,6 +916,30 @@ Some tips and tricks:
+        >>> print(CustomVectorizer().build_analyzer()(u"color colour")) # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+        [...'color', ...'color']
+
+    The same approach could be used for stemming.
```
Maybe mention the number normalization idea here.

vene replied:
The number normalization idea is also performed [in this example](https://github.com/scikit-learn/scikit-learn/blob/master/examples/bicluster/plot_bicluster_newsgroups.py) in our codebase. How about modifying that example to use the same inheritance trick (to avoid duplication of the tokenizing functionality), and then reference it from here?

vene replied:
@jnothman what do you think?

jnothman replied:
Modifying that example and referencing here seems reasonable.

rth committed:
Updated the custom token processing example

vene commented on doc/modules/feature_extraction.rst:
```diff
@@ -937,7 +937,10 @@ Some tips and tricks:
+    The same approach could be used for stemming or normalizing all numbers
+    to a new token, as illustrated in:
+
+     * :ref:`sphx_glr_auto_examples_model_selection_grid_search_text_feature_extraction.py`
```
I think you're referencing the wrong example here

vene commented on doc/modules/feature_extraction.rst:
```diff
@@ -937,7 +937,10 @@ Some tips and tricks:
         [...'color', ...'color']
 
-    The same approach could be used for stemming.
+    The same approach could be used for stemming or normalizing all numbers
```
[...] for other styles of preprocessing; examples include stemming, lemmatization, or normalizing numerical tokens, with the latter illustrated in:

rth committed:
More review comments

vene commented on examples/bicluster/plot_bicluster_newsgroups.py:
```diff
@@ -41,18 +41,21 @@
 
-def number_aware_tokenizer(doc):
-    """ Tokenizer that maps all numeric tokens to a placeholder.
+def numbers_to_paceholder(tokens):
```
number_normalizer?

vene replied:
(paceholder is a typo anyway)

vene commented on examples/bicluster/plot_bicluster_newsgroups.py:
```diff
@@ -41,18 +41,21 @@
+        yield re.sub(r"^\d+$", "#NUMBER", t)
+
+
+class CustomTfidfVectorizer(TfidfVectorizer):
```
class NumberNormalizingVectorizer (or NumberNormalizerTfidfVectorizer, NumberAware(Tfidf)Vectorizer)

vene commented:
as soon as the minor wording fixes are done and we can check the travis/circleci output this is awesome imo!

rth commented:
Thanks for the quick review @vene ! Made the changes, waiting for CI..

rth committed:
Directly call parent method instead of "super"

rth commented:
Also replaced `super(CustomVectorizer, self).build_tokenizer()` by `CountVectorizer.build_tokenizer(self)`. Not sure what's the policy on that...

rth committed:
Revert the last modification relative to "super" call

rth commented:
@vene OK, I reverted the last change regarding `super( )`. CircleCI and Travis passed..

vene commented on doc/modules/feature_extraction.rst:
```diff
@@ -916,6 +916,33 @@ Some tips and tricks:
+        ...
+        >>> class CustomVectorizer(CountVectorizer):
+        ...     def build_tokenizer(self):
+        ...         base = super(CustomVectorizer, self).build_tokenizer()
```
one more thing, sorry!

could you rename this from `base` to `tokenize` or similar? (here as well as in the example)

vene commented on doc/modules/feature_extraction.rst:
```diff
@@ -916,6 +916,33 @@ Some tips and tricks:
+    for other styles of preprocessing; examples include stemming, lemmatization,
+    or normalizing numerical tokens, with the latter illustrated in:
+
+     * :ref:`sphx_glr_auto_examples_bicluster_plot_bicluster_newsgroups.py.py`
```
this link is not working correctly, see rendering [here](https://10998-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/modules/feature_extraction.html)

vene replied:
i think it's because of the double ".py" at the end ;)

rth replied:
Thanks, missed that. Where do you find the URL of the docs rendered by circleci? Circle CI passed, but I'm not sure if it rendered correctly..

vene replied:
see #9082 ;)

rth committed:
More review comments / (retriggering the Circle CI build)

vene commented on examples/bicluster/plot_bicluster_newsgroups.py:
```diff
@@ -41,18 +41,21 @@
-              for token in tokens]
-    return tokens
+    for t in tokens:
+        yield re.sub(r"^\d+$", "#NUMBER", t)
```
This is different behavior tham before. The docstring is no longer accurate: the old version checks if the **first character** is a digit, your regex requires the **entire token** to be made out of digits.

Either update the docstring or (maybe better) replace by

```
return ("#NUMBER" if token[0].isdigit() else token for token in tokens)
```

Come to think of it, even if you want to keep your behaviour, it is probably faster to use


```
return ("#NUMBER" if token.isdigit() else token for token in tokens)
```

vene replied:
On this dataset there seem to be quite a few tokens that start with digits and should probably be discarded, so it might be a good idea to showcase this idea, as well as to keep behavior as close to the old example as possible.

(I'm ok with removing the underscore, that was not really explained in the old example)

rth replied:
Done, reverted to the original behavior. Sorry for all the changes..

ogrisel commented:
LGTM, merging.

rth commented:
Thanks for the review, @ogrisel .

vene commented:
Thanks @rth this is a very nice example!
